Sources/GRPCCodeGen/Internal/TypeName.swift

```diff
@@ -0,0 +1,127 @@
+  /// - Parameter components: A list of components for the type.
+  init(components: [Component]) {
+    precondition(!components.compactMap(\.swift).isEmpty, "TypeName Swift key path cannot be empty")
+    self.components = components
+  }
+
+  /// Creates a new type name with the specified list of Swift path
+  /// components.
+  ///
+  /// - Parameter swiftKeyPath: A list of Swift path components for the type.
+  init(swiftKeyPath: [String]) {
+    precondition(!swiftKeyPath.isEmpty, "TypeName Swift key path cannot be empty")
+    self.init(components: swiftKeyPath.map { .init(swift: $0) })
+  }
+
+  /// A string representation of the fully qualified Swift type name.
+  ///
+  /// For example: `Swift.Int`.
+  var fullyQualifiedSwiftName: String { swiftKeyPathComponents.joined(separator: ".") }
+
+  /// A string representation of the last path component of the Swift
+  /// type name.
+  ///
+  /// For example: `Int`.
+  var shortSwiftName: String { swiftKeyPathComponents.last! }
+
+  /// Returns a type name by appending the specified components to the
+  /// current type name.
+  ///
+  /// In other words, returns a type name for a child type.
+  /// - Precondition: At least one of the components must be non-nil.
+  /// - Parameters:
+  ///   - swiftComponent: The name of the Swift type component.
+  /// - Returns: A new type name.
+  func appending(swiftComponent: String? = nil) -> Self {
+    precondition(swiftComponent != nil, "The Swift name must be non-nil.")
+    let newComponent = Component(swift: swiftComponent)
+    return .init(components: components + [newComponent])
+  }
+
+  /// Returns a type name by removing the last component from the current
+  /// type name.
+  ///
+  /// In other words, returns a type name for the parent type.
+  var parent: TypeName {
+    precondition(components.count >= 1, "Cannot get the parent of a root type")
+    return .init(components: components.dropLast())
+  }
+}
+
+extension TypeName: CustomStringConvertible {
+  var description: String {
+    return fullyQualifiedSwiftName
+  }
+}
+
+extension TypeName {
+  /// Returns the type name for the String type.
+  static var string: Self { .swift("String") }
+
+  /// Returns the type name for the Int type.
+  static var int: Self { .swift("Int") }
+
+  /// Returns a type name for a type with the specified name in the
+  /// Swift module.
+  /// - Parameter name: The name of the type.
+  /// - Returns: A TypeName representing the specified type within the Swift module.
+  static func swift(_ name: String) -> TypeName { TypeName(swiftKeyPath: ["Swift", name]) }
+
+  /// Returns a type name for a type with the specified name in the
+  /// Foundation module.
+  /// - Parameter name: The name of the type.
+  /// - Returns: A TypeName representing the specified type within the Foundation module.
+  static func foundation(_ name: String) -> TypeName {
+    TypeName(swiftKeyPath: ["Foundation", name])
```

Comment:
This extension is from another file - `Builtins.swift`. Should I signal this in any way?

Reply:
No, this is fine.